File: product/src/main/scala/de/leanovate/dose/product/consul/ConsulLookup.scala
package de.leanovate.dose.product.consul

import spray.client.pipelining._
import spray.http.HttpRequest
import spray.httpx.SprayJsonSupport._

import scala.concurrent.Future

object ConsulLookup {

  import de.leanovate.dose.product.Akka._

  val pipeline: HttpRequest => Future[Seq[ServiceNode]] =
    sendReceive ~> unmarshal[Seq[ServiceNode]]

  def lookup(serviceName: String): Future[Seq[ServiceNode]] = {
    val request = Get(s"http://localhost:8500/v1/catalog/service/$serviceName")

    pipeline(request)
  }
}

File: customer/src/main/java/de/leanovate/dose/customer/consul/ConsulLookup.java
package de.leanovate.dose.customer.consul;

import org.springframework.web.client.RestTemplate;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.Arrays;
import java.util.List;

public class ConsulLookup {
    public static List<ServiceNode> lookup(final String serviceName) {

        try {
            final RestTemplate restTemplate = new RestTemplate();
            final ServiceNode[] serviceNodes = restTemplate
                    .getForObject("http://localhost:8500/v1/catalog/service/" + URLEncoder.encode(serviceName, "UTF-8"),
                            ServiceNode[].class);

            return Arrays.asList(serviceNodes);
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }
}
